src/browser/api/window.js

```diff
@@ -1051,25 +1051,95 @@ Window.getParentApplication = function() {
+    // Each successfully loaded script is added below in a `script` property.
+    let payloads = preloads.map((preload, index) => {
+        if (typeof preload !== 'object') {
+            preload = { url: preload };
         }
+        return Object.assign({
+            index,
+            description: `${preload.optional ? 'optional' : 'required'} preload script ${preload.url}`
+        }, preload);
+    });
 
-        fs.readFile(scriptPath, 'utf8', (readError, content) => {
-            if (readError) {
-                callback(new Error('Failed to read the content of the preload script'));
+    payloads.forEach(payload => {
+        cachedFetch(identity.uuid, payload.url, (fetchError, scriptPath) => {
+            if (failed) {
+                return;
+            }
+
+            if (fetchError) {
+                scriptsRemaining -= 1;
+                if (!payload.optional) {
+                    //todo: Abort all other fetches-in-progress before nack'ing
+                    failed = true;
+                    nack(new Error(`Failed to fetch ${payload.description} so no preload scripts will be executed.`));
+                } else if (!scriptsRemaining) {
+                    ack(payloads);
+                }
             } else {
-                callback(null, content);
+                fs.readFile(scriptPath, 'utf8', (readError, script) => {
+                    if (!failed) {
+                        scriptsRemaining -= 1;
+
+                        // todo: remove following workaround when RUN-3162 issue fixed
+                        //BEGIN WORKAROUND (RUN-3162 fetchError null on 404)
+                        if (!readError) {
+                            if (
+                                !script.indexOf('Cannot GET ') || // 404 message (grunt-opefin)
+                                !script.indexOf('<?xml') // 404 message (test_runner)
+                            ) {
+                                if (!payload.optional) {
+                                    failed = true;
+                                    nack(new Error(`Failed to fetch ${payload.description} so no preload scripts will be executed.`));
+                                } else if (!scriptsRemaining) {
+                                    ack(payloads);
+                                }
+                            }
+                        }
+                        //EMD WORKAROUND
+
+                        if (!readError) {
+                            payload.script = script;
+                            if (!scriptsRemaining) {
+                                ack(payloads);
+                            }
+                        } else if (!payload.optional) {
+                            failed = true;
+                            nack(new Error(`Failed to load cached ${payload.description} so no preload scripts will be executed.`));
+                        } else if (!scriptsRemaining) {
+                            ack(payloads);
```

Comment:
Saw 3 repetition of `if (!scriptsRemaining) { ack(payloads); }`, probably, should DRY it?